File: templates/unionfind/union_find.go
package unionfind

type UnionFind struct {
	parent []int
	rank   []int
}

func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
	}
	return &UnionFind{parent: parent, rank: rank}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x])
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(a, b int) bool {
	ra, rb := uf.Find(a), uf.Find(b)
	if ra == rb {
		return false
	}

	if uf.rank[ra] < uf.rank[rb] {
		ra, rb = rb, ra
	}
	uf.parent[rb] = ra
	if uf.rank[ra] == uf.rank[rb] {
		uf.rank[ra]++
	}
	return true
}

File: python/unionfind/unionfind.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Config:
    size: int


class UnionFind(Protocol):

    def size(self) -> int:
        ...

    def component_count(self) -> int:
        ...

    def find(self, value: int) -> tuple[int, bool]:
        ...

    def union(self, left: int, right: int) -> bool:
        ...

    def connected(self, left: int, right: int) -> bool:
        ...

    def component_size(self, value: int) -> tuple[int, bool]:
        ...
